src/Php/PolyfillPackagesProvider.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\Core\Php;
+
+use Nette\Utils\FileSystem;
+use Nette\Utils\Json;
+use Rector\Core\Configuration\Option;
+use Rector\Core\Configuration\Parameter\SimpleParameterProvider;
+use Rector\Core\ValueObject\PolyfillPackage;
+
+final class PolyfillPackagesProvider
+{
+    /**
+     * @var array<PolyfillPackage::*>
+     */
+    private array $cachedPolyfillPackages = [];
+
+    /**
+     * @return array<PolyfillPackage::*>
+     */
+    public function provide(): array
+    {
+        // used in tests mostly
+        if (SimpleParameterProvider::hasParameter(Option::POLYFILL_PACKAGES)) {
+            return SimpleParameterProvider::provideArrayParameter(Option::POLYFILL_PACKAGES);
+        }
+
+        $projectComposerJson = getcwd() . '/composer.json';
+        if (! file_exists($projectComposerJson)) {
+            return [];
+        }
+
+        if ($this->cachedPolyfillPackages !== []) {
+            return $this->cachedPolyfillPackages;
+        }
+
+        $composerContents = FileSystem::read($projectComposerJson);
+        $composerJson = Json::decode($composerContents, Json::FORCE_ARRAY);
+
+        $this->cachedPolyfillPackages = $this->filterPolyfillPackages($composerJson['require'] ?? []);
```

Comment:
this fallback to empty array means the cache never works on no polifill found, that means composer.json will be re-read, that's why nullable for init value comparison is needed instead of compare to non-empty array